network/p2p/gossip/gossip.go

```diff
@@ -176,8 +225,113 @@ func (p *PullGossiper[T, U]) handleResponse(
 		}
 	}
 
-	p.receivedN.Add(float64(len(response.Gossip)))
-	p.receivedBytes.Add(float64(receivedBytes))
+	receivedCountMetric, err := p.metrics.receivedCount.GetMetricWith(p.labels)
+	if err != nil {
+		p.log.Error("failed to get received count metric", zap.Error(err))
+		return
+	}
+
+	receivedBytesMetric, err := p.metrics.receivedBytes.GetMetricWith(p.labels)
+	if err != nil {
+		p.log.Error("failed to get received bytes metric", zap.Error(err))
+		return
+	}
+
+	receivedCountMetric.Add(float64(len(response.Gossip)))
+	receivedBytesMetric.Add(float64(receivedBytes))
+}
+
+// NewPushGossiper returns an instance of PushGossiper
+func NewPushGossiper[T Gossipable](client *p2p.Client, metrics Metrics, maxGossipSize int) *PushGossiper[T] {
+	return &PushGossiper[T]{
+		client:        client,
+		metrics:       metrics,
+		maxGossipSize: maxGossipSize,
+		labels: prometheus.Labels{
+			typeLabel: pushType,
+		},
+		pending: buffer.NewUnboundedDeque[T](0),
+	}
+}
+
+// PushGossiper broadcasts gossip to peers randomly in the network
+type PushGossiper[T Gossipable] struct {
+	client        *p2p.Client
+	metrics       Metrics
+	maxGossipSize int
+
+	labels prometheus.Labels
+
+	lock    sync.Mutex
+	pending buffer.Deque[T]
+}
+
+// Gossip flushes any queued gossipables
+func (p *PushGossiper[T]) Gossip(ctx context.Context) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if p.pending.Len() == 0 {
+		return nil
+	}
+
+	msg := &sdk.PushGossip{
+		Gossip: make([][]byte, 0, p.pending.Len()),
+	}
+
+	sentBytes := 0
+
+	for {
```

Comment:
Rather than breaking at the end I feel like this is a bit cleaner - also rather than calling it `maxGossipSize` I feel like it should be something like `targetGossipSize`
```suggestion
	for sentBytes < p.targetGossipSize  {
```